rfcs/017 - Cluster Data Storage RFC.md

```diff
@@ -0,0 +1,184 @@
+
+Further, while a full network partition between two groups of nodes will result in a "split brain" scenario in an unavoidable fashion, with LogRendezvousHash it is possible to counteract a partial split by way of a request forwarding mechanism. If a node cannot directly reach another node, it should try to ask all other nodes in its list if any of them can reach the node. If so, the first one to respond in the positive becomes the new "router" that all requests to that node are forwarded through, with periodic attempts in the background to remove the extra hop by re-polling. Only if all nodes a node can reach report that they also cannot reach the node in question should it be marked as "down" and assignment of keys shuffled with the new, reduced set. This is fundamentally how the internet routes packets to unknown MAC addresses, and can guard against misconfigured firewalls/routers black hole-ing packets that they shouldn't be, but only works when there is an unaffected third party, so requires at least three separate sub-clusters to work.
+
+For example, let's presume a cluster has been deployed to all three major cloud providers: AWS, Azure, and GCP. Now let's presume that GCP has a network configuration error that makes AWS nodes unreachable to it and vice versa. When the GCP nodes want to get data that was previously assigned to AWS, they will fail to contact the AWS node, then try to ask all other nodes (which would include the other AWS nodes) which ones can reach the AWS node in question. The GCP nodes would report a failure, but the Azure nodes would report a success, and one of those nodes will be chosen effectively at random to be the node that will route to the AWS node. This process would repeat across every GCP node and AWS node combination with an effectively random assignment of Azure nodes to handle the load, which should lead to an approximately even level of traffic distribution between the Azure nodes.
+
+![A partial partition diagram of the paragraph above](./partial_partition.jpg)
+
+This all can be contained in a relatively simple `/ping/<public_ip_address>` control port endpoint. Normally the `/ping/<public_ip_address>` would be directed at `<public_ip_address>`, which should always return `ok` (as long as the cluster secret token is also provided as a header, otherwise `fail`, but that's at an earlier part of the control port implementation). But if a request for `/ping/<public_ip_address>` comes in and `<public_ip_address>` is not the node's own public IP address, then it immediately forwards the request on to the public IP address in the request URL and returns its response or `fail` if it times out (should be a short timeout as well, such as 50ms, but exact threshold depends on real world data here).
```

Comment:
why do we need this instead of just utilizing ICMP? It's more direct, already exposed, and we don't have to write anything to handle the logic on the receiver end - we can just 1 ICMP echo packet directly to the node and if we get a response, we know the node is online.

Reply:
So our definition of "up" and "down" isn't just "is there a working Linux kernel on the other side" but "is the AVM on the other side actually working" which requires both listening on port 4142 and the Tokio event loop not being "stuck" (eg user's code has either hit a bug or slammed the process such that it can't respond to external requests correctly).

Also, since we want to use this for potential forwarding logic, the IP address in the `/ping/` request isn't guaranteed to be the IP address the request is pointed at, and when that's the case, extra logic within the AVM to also potentially update its own status of the node in question should be triggered (if I am being requested to see if another node is down, it is suspect; if my own attempt to ping it fails, I should immediately trigger asking the other nodes if they have seen it, not including the node that triggered this request of me; and if I already have them marked as down, I should just report them as down immediately since I already have my own background Tokio task to periodically check if that node is back online). An ICMP packet isn't capable of all of this downstream work triggering. :)